tensorflow/lite/delegates/webnn/webnn_delegate.cc

```diff
@@ -350,37 +337,49 @@ class Subgraph {
     }
 
     if (any_pointers_changed) {
-      graph_inputs_ = wnn::CreateNamedInputs();
+      graph_inputs_ = emscripten::val::object();
       for (int t : inputs_) {
-        wnn_inputs_[t].resource.arrayBufferView.buffer = context->tensors[t].data.raw;
-        wnn_inputs_[t].resource.arrayBufferView.byteLength = context->tensors[t].bytes;
         std::string name = std::to_string(t);
-        graph_inputs_.Set(name.c_str(), &wnn_inputs_[t]);
+        auto input_size = context->tensors[t].bytes / 4;
+        auto input_data = context->tensors[t].data.f;
+        emscripten::val view{ emscripten::typed_memory_view(input_size, input_data) };
+        auto input = emscripten::val::global("Float32Array").new_(input_size);
+        input.call<void>("set", view);
+        graph_inputs_.set(name, input);
       }
 
-      graph_outputs_ = wnn::CreateNamedOutputs();
+      graph_outputs_ = emscripten::val::object();
       for (int t : outputs_) {
-        wnn_outputs_[t].arrayBufferView.buffer = context->tensors[t].data.raw;
-        wnn_outputs_[t].arrayBufferView.byteLength = context->tensors[t].bytes;
         std::string name = std::to_string(t);
-        graph_outputs_.Set(name.c_str(), &wnn_outputs_[t]);
+        auto output_size = context->tensors[t].bytes / 4;
+        auto output_data = context->tensors[t].data.f;
+        emscripten::val view{emscripten::typed_memory_view(output_size, output_data)};
+        auto output = emscripten::val::global("Float32Array").new_(output_size);
+        output.call<void>("set", view);
+        graph_outputs_.set(name, output);
       }
     }
 
-    wnn_graph_.Compute(graph_inputs_, graph_outputs_);
+    wnn_graph_.call<void>("compute", graph_inputs_, graph_outputs_);
+    // Copy output data from JS to TFLite output tensors' buffer.
+    for (int t : outputs_) {
+      auto output_tmp = emscripten::convertJSArrayToNumberVector<float>(graph_outputs_[std::to_string(t)]);
+      std::memcpy(context->tensors[t].data.f, output_tmp.data(), output_tmp.size()*sizeof(float));
```

Comment:
Does it really need to memory copy the data? You created the Float32Array at line 356 with data pointer.

Reply:
> Does it really need to memory copy the data?

Yes, it does. I should remove the code at line 355/356/358 which doesn't make the data map work, since the `view` is not a member of the `output`.

Reply:
If you could use view in line 356, you probably won't need to do memory copy here.